source/simulator/freesurface.cc

```diff
@@ -292,20 +315,30 @@ namespace aspect
               cell_vector = 0;
               cell_matrix = 0;
               for (unsigned int point=0; point<n_face_q_points; ++point)
-                for (unsigned int i=0; i<dofs_per_cell; ++i)
-                  {
-                    for (unsigned int j=0; j<dofs_per_cell; ++j)
-                      {
-                        cell_matrix(i,j) += (fs_fe_face_values[extract_vel].value(j,point) *
-                                             fs_fe_face_values[extract_vel].value(i,point) ) *
-                                            fs_fe_face_values.JxW(point);
-                      }
-
-                    cell_vector(i) += (fs_fe_face_values[extract_vel].value(i,point) *
-                                       fs_fe_face_values.normal_vector(point) ) *
-                                      (velocity_values[point]*fs_fe_face_values.normal_vector(point)) *
-                                      fs_fe_face_values.JxW(point);
-                  }
+                {
+                  //Select the direction onto which to project the velocity solution
+                  Tensor<1,dim> direction;
+                  if ( advection_direction == SurfaceAdvection::normal ) //project onto normal vector
+                    direction = fs_fe_face_values.normal_vector(point);
+                  else if ( advection_direction == SurfaceAdvection::gravity ) //project onto local gravity
+                    direction = sim.gravity_model->gravity_vector( fs_fe_face_values.quadrature_point(point) );
+                  else ExcInternalError();
```

Comment:
This last line creates an unnamed ExcInternalError object, destroys it again, and goes on with life :-)

You'll need something like
```
  AssertThrow (false, ExcInternalError());
```